### PR TITLE
Allow std::vector<std::unique_ptr<T>> be produced, and be read with edm::View<T>

### LINKED ISSUE #27277: Dictionaries for new DQM Dataformat
Dear framework experts,

the new DataFormat for DQM proposed in #26603 has evolved a bit over the last weeks and we got to actually using it as a product. The latest version can be found here:
https://github.com/schneiml/cmssw/blob/726d7c14e5c71977923c674674f64f84c041042f/DataFormats/Histograms/interface/MonitorElementCollection.h

Now edm tells me
```
No data dictionary found for the following classes:

  std::unique_ptr<MonitorElementData,std::default_delete<MonitorElementData> >
```

but trying to add an entry to `classes_def.xml` leads to the error
```
>> Building LCG reflex dict from header file src/DataFormats/Histograms/src/classes.h
Error: It is not necessary to explicitly select class unique_ptr<MonitorElementData,default_delete<MonitorElementData> >. I/O is supported for it transparently.
```

How to proceed? 

More specifically, which constraints apply to persistent and non-persistent products? For persistence, it seems copying needs to be supported, which makes the use of `unique_ptr` rather hard (and obviously the locks would need to be made transient somehow). There seem to be less constraints for non-persistent products, but still something is wrong here...

I can't see anything in my code that would be mentioned in the relevant TWiki page: https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Restrictions_on_the_package_defi 

For context, the branch https://github.com/cms-sw/cmssw/compare/master...schneiml:dqm-new-dqmstore-on-CMSSW_11_0_0_pre1 should apply cleanly on pre1 and contains some test code in the `DQMServices/Demo` package. However, not all of CMSSW compiles at the moment: only `DataFormats`, `DQMServices` and `DQM`.

@andrius-k @fioriNTU FYI

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

Fixes #27277. The complaint on missing dictionary for `std::unique_ptr<T>` came from the support code for `edm::View<T>`. The minimal fix was to specialize the `edm::detail::getMemberType` for `std::vector<std::unique_ptr<T>>`. That alone allows the `std::vector<std::unique_ptr<T>>` be produced by an EDProducer, and written to a file. Reading such a product from a file appears to currently work only if `T` has a dictionary, i.e. it does not work e.g. for fundamental types.

After that it was natural to complete the support for `edm::View`, i.e. with this PR `edm::View<T>` can be used to read `std::vector<std::unique_ptr<T>>`.

In addition the "friendly name" transformation is improved to ignore the `std::default_delete<T>` from the full type name of `std::unique_ptr1` (among some other improvements in the parsing).

#### PR validation:

Code compiles, unit tests run.

### CODE REVIEW

File: FWCore/Utilities/src/FriendlyName.cc [hunk 1]
```diff
@@ -16,6 +16,11 @@
 // templated class.  Up front we'd register a class name to a transformation function (which would probably take a std::vector<std::string> which holds
 // the results of the node transformations)
 
+namespace {
+  constexpr bool debug = false;
+  std::string prefix; // used only if debug == true
```
Comment (Dr15Jones) [inline FWCore/Utilities/src/FriendlyName.cc]: This is likely to be flagged by the static analyzer as not being thread safe

**Comment by makortel** [comment]
Maybe some day I'll remember to run the code formatting before a PR...

File: FWCore/Utilities/src/FriendlyName.cc (hunk 1 above)
Comment (makortel) [reply FWCore/Utilities/src/FriendlyName.cc]: Good point. Should I then change the `debug` above to a macro and hide the global `prefix` by default? Or do you have better ideas?

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by davidlange6** [comment]
any chance this PR is causing the IB failures that started recently?

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc700/CMSSW_11_0_DEVEL_X_2019-08-28-2300/pyRelValMatrixLogs/run/4.23_ValSkim2011A+ValSkim2011A+HLTDSKIM+RECODR1reHLT+HARVESTDR1reHLT/step2_ValSkim2011A+ValSkim2011A+HLTDSKIM+RECODR1reHLT+HARVESTDR1reHLT.log

An exception of category 'UnimplementedFeature' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing input source of type PoolSource
Exception Message:
Cannot change friendly class name algorithm without more development work
to update BranchIDLists and ThinnedAssociationsHelper.  Contact the framework group.
----- End Fatal Exception -------------------------------------------------

**Comment by makortel** [comment]
It is, @Dr15Jones already notified me and I'm, working on a fix.

**Comment by makortel** [comment]
The fix is in https://github.com/cms-sw/cmssw/pull/27899

**Pull request merged**